wip: github copilot extension (#20208)
docker/docs: content/copilot/index.md

keywords: Docker, GitHub Copilot, extension, Visual Studio Code, chat, ai, containerization
---

> **Early Access**
>
> The Docker for GitHub Copilot extension is a part of the [Limited Public Beta](https://github.blog/changelog/2024-05-21-copilot-extensions-now-in-limited-public-beta/)
> for GitHub Copilot Extensions. Only members of the beta program are eligible to use this extension.
{ .restricted }

The Docker Extension for GitHub Copilot integrates Docker's capabilities with
GitHub Copilot, providing assistance with containerizing applications and
generating Docker assets. This extension helps you streamline Docker-related
tasks wherever GitHub Copilot Chat is available.

The key features are:

- Ask questions and receive responses about containerization in any context
  when GitHub Copilot Chat is available, such as on GitHub.com and in Visual Studio Code.
- Automatically generate Dockerfiles, Docker Compose files, and .dockerignore
  files for a project.
- Open pull requests with generated Docker assets directly from the chat
  interface.
- Get summaries of project vulnerabilities from Docker Scout and receive next
  steps via CLI.

## Supported languages

> Currently, the Docker Extension for GitHub Copilot supports the following

Reviewer: [nit] Possibly remove to align with style guide:
```suggestion
The Docker Extension for GitHub Copilot supports the following
```

programming languages:

- Dockerfile
- Java
- JavaScript
- Python
- TypeScript

## Get Started

Here's how to get started with the Docker Extension for GitHub Copilot.

### Prerequisites

Before you start, ensure that:

- You have been granted access to GitHub's [limited public beta program](https://github.blog/changelog/2024-05-21-copilot-extensions-now-in-limited-public-beta/).
- The [Docker for GitHub Copilot](https://github.com/marketplace/docker-for-github-copilot)
  app is installed in your GitHub organization.

### Setup

Select where you want to use the Docker Extension for GitHub Copilot:

{{< tabs >}}
{{< tab name="Editor or IDE" >}}

For instructions on how to use the Docker Extension for GitHub Copilot in
your editor, see:

- [Visual Studio Code](https://docs.github.com/en/copilot/github-copilot-chat/copilot-chat-in-ides/using-github-copilot-chat-in-your-ide?tool=vscode)
- [Visual Studio](https://docs.github.com/en/copilot/github-copilot-chat/copilot-chat-in-ides/using-github-copilot-chat-in-your-ide?tool=visualstudio)

> - [JetBrains IDEs](https://docs.github.com/en/copilot/github-copilot-chat/copilot-chat-in-ides/using-github-copilot-chat-in-your-ide?tool=jetbrains)

Reviewer 2: Let's remove the JetBrains note for now.

It looks like it's limited to VS, VS Code and github.com right now,  based on their docs
https://docs.github.com/en/copilot/github-copilot-chat/github-copilot-extensions/about-github-copilot-extensions

3. Select the Copilot logo in the site menu to open the chat interface.

   ![Copilot chat button](/copilot/copilot-button.png?w=400px)

4. Interact with the Docker agent by tagging `@docker`.

## Example prompts

Here are some examples of the types of questions you can ask the Docker agent:

### Ask general Docker questions

You can ask general question about Docker. For example:

- `@docker what is a Dockerfile`
- `@docker how do I build a Docker image`
- `@docker how do I run a Docker container`

### Ask questions about your project

You can ask questions about your project, such as:

- `@docker what is the best way to Dockerize this project`
- `@docker can you help me find vulnerabilities in my project?`

The Docker agent will analyze your project, generate the necessary files, and,
if applicable, offer to raise a pull request with the necessary Docker assets.

## Performing actions on your behalf

Before the agent performs any actions on your behalf, such as opening a pull
request for you, you're prompted to provide your consent to allow the
operation. You can always roll back or back out of the changes.

![Copilot action prompt](/copilot/copilot-action-prompt.png?w=400px)

> In the event that the agent encounters and error, for example during PR

Reviewer: Small typo: 
```suggestion
In the event that the agent encounters an error, for example during PR
```

---
title: Docker for GitHub Copilot
description: Learn how to use the Docker Extension for GitHub Copilot to streamline Docker-related tasks directly within GitHub and Visual Studio Code.
keywords: Docker, GitHub Copilot, extension, Visual Studio Code, chat, ai, containerization
sitemap: false
---

> **Early Access**
>

> > The Docker for GitHub Copilot extension is in [Early Access](../../release-lifecycle.md).

Reviewer 2: This is tricky because it's part of GitHub's limited public beta. Not sure if we want to mention that. We are beholden to their timelines on public preview etc.

Author: Yeah let's mention that! I will update.

{ .restricted }

The Docker Extension for GitHub Copilot integrates Docker's capabilities with
GitHub Copilot, providing assistance with containerizing applications and
generating Docker assets. This extension helps you streamline Docker-related
tasks directly within GitHub and Visual Studio Code.

The key features are:

- Ask questions and receive responses about containerization directly within
  GitHub or Visual Studio Code.
- Automatically generate Dockerfiles, Docker Compose files, and .dockerignore
  files for a project, using [`docker init`](../reference/cli/docker/init.md).
- Open pull requests with generated Docker assets directly from the chat
  interface.
- Get summaries of project vulnerabilities from Docker Scout and receive next
  steps via CLI.

## Supported languages

Currently, the Docker Extension for GitHub Copilot supports the following
programming languages:

- Dockerfile
- Java
- JavaScript
- Python
- TypeScript

## Get Started

Here's how to get started with the Docker Extension for GitHub Copilot.

### Prerequisites

Before you start, ensure that:

> - You have been granted access to the private beta program.

Reviewer 2: We may want to link to https://github.blog/2024-05-21-introducing-github-copilot-extensions/? This has details that explain that it's currently a closed beta etc.

- The [Docker for GitHub Copilot](https://github.com/marketplace/docker-for-github-copilot)
  app is installed in your organization.

### Setup

Select where you want to use the Docker Extension for GitHub Copilot:

- **GitHub:**

  1. Navigate to your repository.
  2. Enable the Docker Extension for GitHub Copilot via the GitHub Marketplace.

> - **Visual Studio Code:**

Reviewer 2: This is also available in other places where Copilot is available (e.g. Codespaces, JetBrains IDEs) - https://docs.github.com/en/copilot/github-copilot-chat/copilot-chat-in-ides/using-github-copilot-chat-in-your-ide

Author: yeah I saw that in the docs. I wasn't sure it would be supported in all of those, but I assume it is then! There is no official vim/neovim copilot chat - but there is a copilot.vim plugin. I assume that means we don't "support" vim/nvim until there is an officially supported chat plugin. But only the other three IDEs

  1. Open Visual Studio Code.

>   2. Install the Docker Extension for GitHub Copilot from the Visual Studio Code Marketplace.

Reviewer 2: This isn't quite right - the user will need to install the Copilot extension https://marketplace.visualstudio.com/items?itemName=GitHub.copilot and then sign in with their GitHub account that has organizational access to an org where the Docker extension is installed.

Then they'll just have `/docker` slash prompts in their chat context.

Author: Thank you, I will try it out and update accordingly

> ---

Reviewer 2: General comment - I couldn't see where this comment was linked from in the preview of the docs. 

Where is it linked from :) ?

Author: I've added a navigation entry for this page - I wasn't sure if we wanted to have it listed in the navigation, or not. I also removed the `sitemap: false` metadata, which means that this page will be indexed by search engines and searchable.

title: Docker for GitHub Copilot
description: Learn how to use the Docker Extension for GitHub Copilot to streamline Docker-related tasks directly within GitHub and Visual Studio Code.
keywords: Docker, GitHub Copilot, extension, Visual Studio Code, chat, ai, containerization
sitemap: false
---

> **Early Access**
>
> The Docker for GitHub Copilot extension is a part of the [Limited Public Beta](https://github.blog/changelog/2024-05-21-copilot-extensions-now-in-limited-public-beta/)
> for GitHub Copilot Extensions. Only members of the beta program are eligible to use this extension.
{ .restricted }

The Docker Extension for GitHub Copilot integrates Docker's capabilities with
GitHub Copilot, providing assistance with containerizing applications and
generating Docker assets. This extension helps you streamline Docker-related
tasks directly within GitHub and Visual Studio Code.

The key features are:

> - Ask questions and receive responses about containerization directly within

Reviewer 2: ```suggestion
- Ask questions and receive responses about containerization in any context when GitHub Copilot Chat is available (e.g. github.com, Visual Studio Code)
```

>   GitHub or Visual Studio Code.

Reviewer 2: ```suggestion
  
```

- Automatically generate Dockerfiles, Docker Compose files, and .dockerignore

>   files for a project, using [`docker init`](../reference/cli/docker/init.md).

Reviewer 2: I'm not sure if we want to say this uses Docker Init quite yet. It may be more confusing because we only support a subset of the supported languages/it's sort of an implementation detail.

- Open pull requests with generated Docker assets directly from the chat
  interface.
- Get summaries of project vulnerabilities from Docker Scout and receive next
  steps via CLI.

## Supported languages

Currently, the Docker Extension for GitHub Copilot supports the following
programming languages:

- Dockerfile
- Java
- JavaScript
- Python
- TypeScript

## Get Started

Here's how to get started with the Docker Extension for GitHub Copilot.

### Prerequisites

Before you start, ensure that:

> - You have been granted access to the [private beta program](https://github.blog/changelog/2024-05-21-copilot-extensions-now-in-limited-public-beta/).

Reviewer 2: ```suggestion
- You have been granted access to the [private beta program](https://github.blog/changelog/2024-05-21-copilot-extensions-now-in-limited-public-beta/) by GitHub.
```

### GitHub.com

To use the Docker Extension in the GitHub web interface:

1. Go to [github.com](https://github.com/) and sign in to your account.
2. Go to any repository.
3. Select the Copilot logo in the site menu to open the chat interface.

   ![Copilot chat button](/copilot/copilot-button.png?w=400px)

4. Interact with the Docker agent by tagging `@docker`.

![Docker Copilot extension on GitHub](/copilot/CopilotDemo.gif)

## Example prompts

Here are some examples of the types of questions you can ask the Docker agent:

### Ask general Docker questions

You can ask general question about Docker. For example:

- `@docker what is a Dockerfile`
- `@docker how do I build a Docker image`
- `@docker how do I run a Docker container`

### Ask questions about your project

You can ask questions about your project, such as:

- `@docker what is the best way to Dockerize this project`
- `@docker how do I create a Dockerfile for this project`
- `@docker how do I create a Docker Compose file for this project`

The Docker agent will outline the steps for you to follow, and where possible,
automatically raise a pull request with the necessary Docker assets.

> ### Update Docker assets and code

Reviewer 2: Woah, did you try these out (here down)? Did they work?!

Author: I did try them - but only some. The agent started having technical difficulties as I was going through these prompts. I will run through them once more.
